pkg/driver/driver_test.go

```diff
@@ -124,6 +132,88 @@ func TestDriverStop(t *testing.T) {
 	driver.Stop()
 }
 
+// TestControllerOnlyAffectsMounterCreation verifies that when CSI_CONTROLLER_ONLY is true,
+// the driver skips mounter initialization and thus has a nil NodeServer; otherwise it creates one.
+func TestControllerOnlyAffectsMounterCreation(t *testing.T) {
+	t.Parallel()
+
+	// Save and restore env vars
+	originalControllerOnly := os.Getenv("CSI_CONTROLLER_ONLY")
+	originalEndpointURL := os.Getenv(envprovider.EnvEndpointURL)
+	defer func() {
+		if originalControllerOnly == "" {
+			_ = os.Unsetenv("CSI_CONTROLLER_ONLY")
+		} else {
+			_ = os.Setenv("CSI_CONTROLLER_ONLY", originalControllerOnly)
+		}
+		if originalEndpointURL == "" {
+			_ = os.Unsetenv(envprovider.EnvEndpointURL)
+		} else {
+			_ = os.Setenv(envprovider.EnvEndpointURL, originalEndpointURL)
+		}
+		// restore seams
+		driver.InClusterConfigTestHook(nil)
+		driver.KubeClientForConfigTestHook(nil)
+		driver.KubernetesVersionTestHook(nil)
+		driver.NewSystemdMounterTestHook(nil)
+	}()
+
+	// Provide required env for NewDriver validation
+	_ = os.Setenv(envprovider.EnvEndpointURL, "http://s3.example.com:8000")
+
+	// Hook the external dependencies so NewDriver is unit-testable
+	driver.InClusterConfigTestHook(func() (*rest.Config, error) {
+		// return a dummy config to allow client creation
+		return &rest.Config{Host: "http://localhost"}, nil
+	})
+	driver.KubeClientForConfigTestHook(func(*rest.Config) (kubernetes.Interface, error) {
+		// return a fake clientset
+		return fake.NewSimpleClientset(), nil
+	})
+	driver.KubernetesVersionTestHook(func(_ kubernetes.Interface) (string, error) {
+		return "v1.30.0", nil
+	})
+	// Avoid exercising systemd mounter in tests; just ensure code path completes
+	driver.NewSystemdMounterTestHook(func(_ *credentialprovider.Provider, _ string, _ string) (mounterpkg.Mounter, error) {
+		return &noopMounter{}, nil
+	})
+
+	// 1) controller-only path: NodeServer should be nil
+	_ = os.Setenv("CSI_CONTROLLER_ONLY", "true")
+	d1, err := driver.NewDriver("unix:///tmp/test.sock", "mpv", "node-1")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if d1.NodeServer == nil {
+		// expected
+	} else {
+		t.Fatalf("expected NodeServer to be nil in controller-only mode")
+	}
+
+	// 2) node path: NodeServer should be non-nil; force systemd path by clearing pod mounter env
+	_ = os.Setenv("CSI_CONTROLLER_ONLY", "false")
+	_ = os.Unsetenv("MOUNTER_KIND")
+	d2, err := driver.NewDriver("unix:///tmp/test.sock", "mpv", "node-2")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if d2.NodeServer == nil {
+		t.Fatalf("expected NodeServer to be initialized when not controller-only")
+	}
+}
+
+// minimal noop mounter for testing the systemd branch without side effects
+type noopMounter struct{}
```

Comment:
minor: This is defined after use.

Reply:
Thanks, I will fix this as this whole thig will change